### PR TITLE
iOS a11y: Implement strong down weak up reference management

### PR DESCRIPTION
**Pull request by yjbanov**
Fixes https://github.com/flutter/flutter/issues/13701

The crash was caused by a use-after-free bug. `SemanticsObjectContainer` uses a raw pointer to point to `_semanticsObject`. When it returns the pointer to iOS' a11y framework it does not bump the ref count. Similarly, `SemanticsObject` has a raw `std::vector<SemanticsObject*> _children`, and does not bump the ref count when returning pointers to iOS. When the semantics tree is updated and `AccessibilityBridge::UpdateSemantics` clears the `objects_` dictionary, the ref count hits zero and the semantics objects are freed. In the mean time, iOS continues to assume that the objects are live.

/cc @goderbauer @chinmaygarde

### CODE REVIEW

**Comment by yjbanov** [comment]
Hmm, I do not understand the Travis failure. It seems to tell me that I should format a file that I didn't touch :/

**Comment by goderbauer** [comment]
Looks like the engine repo is red on travis for master, so that failure was probably not introduced by you: https://travis-ci.org/flutter/engine/builds/334200934

**Comment by cbracken** [comment]
+1 to @chinmaygarde's comment. First thing to do is clarify who's responsible for object lifetime, then clean up what's there and stop relying on just the shared dict for ref-count. Everyone using an NSObject should manage their own inc/dec of the refcount.

**Comment by yjbanov** [comment]
PTAL. In this new version I made `parent` a weak pointer, and `children` a strong ref-counted pointer. On top of that, I null out children's `parent` pointer eagerly in the parent's destructor because there's no guarantee about when the child's destructor is called (iOS may continue holding references to children after we cleared them out). A similar relationship is established between `SemanticsObjectContainer` and `SemanticsObject`. After this change, `SemanticsObject` only holds strong references to _it's_ children. So strong references always point down the tree, and weak references point up the tree and cleared out eagerly.

**Comment by yjbanov** [comment]
Update: I can no longer vouch for existence of a leak. It seems I have been reading the Instruments UI wrong. Many of the objects listed in it have ref count == 0.

File: shell/platform/darwin/ios/framework/Source/accessibility_bridge.mm [hunk 1]
```diff
@@ -40,14 +40,13 @@
...
 
-bool GeometryComparator(SemanticsObject* a, SemanticsObject* b) {
-  // Should a go before b?
-  CGRect rectA = [a accessibilityFrame];
-  CGRect rectB = [b accessibilityFrame];
-  CGFloat top = rectA.origin.y - rectB.origin.y;
-  if (top == 0.0)
-    return rectA.origin.x - rectB.origin.x < 0.0;
-  return top < 0.0;
+NSComparisonResult intToComparisonResult(int32_t value) {
```
Comment (cbracken) [inline shell/platform/darwin/ios/framework/Source/accessibility_bridge.mm]: `IntToComparisonResult` for static funcs.

**Pull request merged**